python/cuxfilter/charts/core/core_chart.py

```diff
@@ -70,6 +73,22 @@ def height(self, value):
     def library_specific_params(self):
         return self._library_specific_params
 
+    @property
+    def x_dtype(self):
+        if isinstance(self.source, ColumnDataSource):
+            return self.source.data[self.data_x_axis].dtype
+        elif isinstance(self.source, (cudf.DataFrame, dask_cudf.DataFrame)):
+            return self.source[self.x].dtype
+        return None
```

Comment:
No action here but comment: it seems like the best place to improve this is upstream, i.e. see if there is away for  CDS columns to be cudf series, or cupy arrays. It's possible a new (sub?) type to cope with deferred evaluation or device/host movement would be needed. Also the lib dependencies could only be optional, used if passed in but not otherwise imported. Maybe that's more trouble than it's worth but if you think it's worth exploring I'd say open a Bokeh issue to discuss.

Reply:
Have made a note of it, will try exploring a bit into this, and create a separate issue on bokeh